enhancements/installer/openstack-no-fip-install.md

```diff
@@ -0,0 +1,99 @@
+authors:
+  - "@egarcia"
+  - "@adduarte"
+reviewers:
+  - "@mandre"
+  - "@pierre"
+approvers:
+  - "@mandre"
+  - "@pierre"
+creation-date: 2020-5-26
+last-updated: 2020-5-26
+status: implementable
+---
+
+# OpenStack BYO External Connectivity
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+Numerous customers have requested the ability to run the OpenShift on OpenStack IPI installer without using Floating IP addresses.
+In the installer, we use floating IPs as a way to set up exeternal connectivity to the OpenShift cluster. Depending on how the OpenStack cluster
+is set up, the use of floating IPs may be impossible or very inconvenient. In order to support these customers, we intend to support installations
+on these types of OpenStack clusters.
+
+
+### Non-Goals
+
+This enhancement does not intend to provide any guidance or features that go further than allowing customers to deploy withoug Floating IPs. Operations like attaching loadbalancers to the entrypoints, or any day 2 operations are beyond the scope of this enhancement. 
+
+## Proposal
+
+In order to implement this feature fully, the following changes must be made:
+    1. Make `externalNetwork` an optional argument in the install-config
+    2. Disable `lbFloatingIP` when no external network is passed
+
+### User Stories
+
+#### Provider Networks
+
+As an enterprise OpenStack cluster administrator, I want to use the IPI installer but my cluster does not support Floating IPs. I need a way to install the OpenShift IPI installer on provider networks instead. I may also need to clear IP addresses I use on the provider network through an internal audit.
+
+##### Example Usage
+
+Using the proposed feature, the user would not set `externalNetwork` or `lbFloatingIP`. Then, they will set`machinesSubnet` to the ID of a provider network that they want to install the cluster onto. They will also set custom ports IPs for the `apiVP` and `ingressVIP` to ensure the IP they take is available. Here is an example install config:
+
+```yaml
+apiVersion: v1
+baseDomain: example.com
+metadata:
+  name: test-cluster
+platform:
+  openstack:
+    cloud: mycloud
+    computeFlavor: m1.s2.xlarge
+    externalNetwork:
+    lbFloatingIP:
+    machinesSubnet: fa806b2f-abcd-4bce-b9db-124bc64209bf
+    apiVIP: `10.0.1.19`
+    ingressVIP: `10.0.1.20`
+pullSecret: '{"auths": ...}'
+sshKey: ssh-ed25519 AAAA...
+```
+
+#### No Public External Access
+
+As an administrator with an on-prem OpenStack cluster, I want to install OpenShift in a way that does not make it available to the public internet. Instead, I would prefer to install the cluster without any external connectivity, and attach my own loadbalancers to the cluster entrypoints post install.
```

Comment:
Depending on how the OpenStack cloud is setup, the "external" connectivity might already point to a private subnet.

I think we should rephrase to highlight the desire to not expose API and Ingress endpoints to an external network, either public or private, and instead rely on a customer owned infrastructure (LB, FW) to provide access to the cluster.

In this use case, someone could provision a VM in a tenant network and use it to run the installer. No endpoints would be exposed outside of the specified `machineSubnet`.

Reply:
+1 , the reason of FIP is the installer running on machine is not able to access the deployed one
so we need a FIP, but actually it's not needed if customer provide network access from machine running installer to the new spawned machines
and after cluster is up the connectivity depends on customer setup and configuration as well

Reply:
Gotcha, so in order to do this cleanly, we should allow `externalNetwork` and `lbFloatingIP` to be unset even if `machinesSubnet` is not set

Reply:
No, we should make `externalNetwork` and  `lbFloatingIP` optional only when `machinesSubnet` is set.
See my comment at https://github.com/openshift/enhancements/pull/342#discussion_r431627365

Reply:
I saw your comment, but if we are installing from a node within the cluster, and we don't want any external connectivity, why should we have to use a custom subnet? I think it is confusing to enforce that.

Reply:
Because if you let the installer create the networks, that means that the node running the installer is on a different subnet and doesn't have L3 connectivity to the node subnet.

Instead, if you pre-create a subnet, provision a node on it to run the installer, and pass this subnet ID as the `machineSubnet` then the installation will work as expected.

Reply:
Ahh ok, I follow now